frontend/catalyst/utils/wrapper.cpp

```diff
@@ -1,15 +1,186 @@
+    auto etypes = result_desc.attr("_etypes_");
+    auto sizes = result_desc.attr("_sizes_");
+
+    size_t memref_len = ranks.attr("__len__")().cast<size_t>();
+    size_t offset = 0;
+
+    char *memref_array_bytes = (char *)(memref_array_ptr);
+
+    for (size_t idx = 0; idx < memref_len; idx++) {
+        unsigned int rank_i = ranks.attr("__getitem__")(idx).cast<unsigned int>();
+        char *memref_i_beginning = memref_array_bytes + offset;
+        offset += memref_size_based_on_rank(rank_i);
+
+        struct memref_beginning_t *memref = (struct memref_beginning_t *)memref_i_beginning;
+        bool is_in_rt_heap = f_transfer_ptr(memref->allocated);
+
+        if (!is_in_rt_heap) {
+            // This case is guaranteed by the compiler to be the following:
+            // when an input tensor is sent to as an output as well.
+            //
+            // It is guaranteed by the use of the flag --cp-global-memref
+            //
+            // Use the numpy_arrays dictionary which sets up the following map:
+            // integer (memory address) -> py::object (numpy array)
+            auto array_object = numpy_arrays.attr("__getitem__")((size_t)memref->allocated);
+            returns.append(array_object);
+            continue;
+        }
+
+        const npy_intp *dims = npy_get_dimensions(memref_i_beginning, rank_i);
+
+        size_t element_size = sizes.attr("__getitem__")(idx).cast<size_t>();
+        const npy_intp *strides = npy_get_strides(memref_i_beginning, element_size, rank_i);
+
+        auto etype_i = etypes.attr("__getitem__")(idx);
+        PyArray_Descr *descr = PyArray_DescrFromTypeObject(etype_i.ptr());
+        if (!descr) {
+            throw std::runtime_error("PyArray_Descr failed.");
+        }
+
+        void *aligned = memref->aligned;
+        PyObject *new_array =
+            PyArray_NewFromDescr(&PyArray_Type, descr, rank_i, dims, strides, aligned, 0, NULL);
+        if (!new_array) {
+            throw std::runtime_error("PyArray_NewFromDescr failed.");
+        }
+
+        PyObject *capsule =
+            PyCapsule_New(memref->allocated, NULL, (PyCapsule_Destructor)&free_wrap);
+        if (!capsule) {
+            throw std::runtime_error("PyCapsule_New failed.");
+        }
+
+        int retval = PyArray_SetBaseObject((PyArrayObject *)new_array, capsule);
+        bool success = 0 == retval;
+        if (!success) {
+            throw std::runtime_error("PyArray_SetBaseObject failed.");
+        }
+
+        returns.append(new_array);
+
+        // Now we insert the array into the dictionary.
+        // This is to take into account the possibility
+        // of returning more than 1 array with the same buffer.
+        // Although, it looks like it won't happen in reality.
+        PyObject *pyLong = PyLong_FromLong((size_t)memref->allocated);
+        if (!pyLong) {
+            throw std::runtime_error("PyLong_FromLong failed.");
+        }
+
+        numpy_arrays[pyLong] = new_array;
```

Comment:
Could you explain this part a little more, what dictionary are we inserting the array into, and how does it help with arrays that share the same buffer?

Reply:
The idea is that for arrays which share the same buffer, the runtime has dropped the `allocated` pointer once, therefore it won't appear again. This means that `_mlir_memory_transfer` will return false the second time. But the buffer has already been converted to an array. So the idea is that all buffers that have already been converted into arrays are indexed based on the buffer in this dictionary. Then we can always look them up in this dictionary if `_mlir_memory_transfer` returns false.

Reply:
Thank you that makes a lot of sense now! 
Could you include this case in the comment on L112-118 then, and update this comment to say what dictionary we are inserting the array into? :)